File: backend/app/services/network_metric_service.py
from app.models.network_metric import NetworkMetric

from app.repositories.asset_repository import AssetRepository
from app.repositories.network_interface_repository import (
    NetworkInterfaceRepository,
)
from app.repositories.network_metric_repository import (
    NetworkMetricRepository,
)


class NetworkMetricService:

    @staticmethod
    def create_metric(db, data):

        # Find Asset
        asset = AssetRepository.get_by_hostname(
            db,
            data.hostname,
        )

        if not asset:
            raise ValueError("Asset not found.")

        # Find Interface
        interface = (
            NetworkInterfaceRepository.get_by_name(
                db=db,
                asset_id=asset.id,
                interface_name=data.interface_name,
            )
        )

        if not interface:
            raise ValueError("Network interface not found.")

        metric = NetworkMetric(
            asset_id=asset.id,
            interface_id=interface.id,
            bytes_sent=data.bytes_sent,
            bytes_received=data.bytes_received,
            packets_sent=data.packets_sent,
            packets_received=data.packets_received,
            upload_speed=data.upload_speed,
            download_speed=data.download_speed,
        )

        return NetworkMetricRepository.create(
            db,
            metric,
        )

    @staticmethod
    def get_asset_metrics(
        db,
        asset_id,
        page,
        size,
    ):
        return NetworkMetricRepository.get_by_asset(
            db,
            asset_id,
            page,
            size,
        )

File: frontend/src/services/networkMetricService.ts
import api from "../api/axios";

export interface NetworkMetric {
  id: string;
  asset_id: string;
  interface_id: string;

  bytes_sent: number;
  bytes_received: number;

  packets_sent: number;
  packets_received: number;

  upload_speed: number;
  download_speed: number;

  created_at: string;
}

interface NetworkMetricResponse {
  items: NetworkMetric[];
  total: number;
  page: number;
  size: number;
}

export async function getLatestNetworkMetrics(
  limit = 20
): Promise<NetworkMetric[]> {
  const response = await api.get(
    "/api/v1/network/metrics/latest",
    {
      params: { limit },
    }
  );

  return response.data.items;
}

export async function getAssetNetworkMetrics(
  assetId: string
): Promise<NetworkMetric[]> {
  const response = await api.get(
    `/api/v1/network/metrics/${assetId}`,
    {
      params: {
        page: 1,
        size: 100,
      },
    }
  );

  return response.data.items;
}